ui.af.apps/src/main/content/jcr_root/apps/core/fd/af-clientlibs/core-forms-components-runtime-all/js/datalayer.js

```diff
@@ -0,0 +1,151 @@
+        ABANDON: "Abandoned forms",
+        HELP: "Help views",
+        TIMESPENT: "Time Spent"
+    }
+
+    const getFormName = (dataLayerContent) => {
+        formContainer = Object.keys(dataLayerContent.page)[0];
+        const formName = dataLayerContent.page[formContainer][DC_TITLE];
+        let formPath = dataLayerContent.page[formContainer][REPO_PATH];
+        formPath = formPath.substring(0,formPath.lastIndexOf('/'));
+        return formName + '(' + formPath + ')';
+    }
+    
+    const getCurrentFieldData = (fieldId, dataLayerContent) => {
+        return dataLayerContent.component[fieldId];
+    }
+    
+    const getFieldPanelTitle = (fieldData, dataLayerContent) => {
+        if(dataLayerContent.component[fieldData[PARENT_ID]] === undefined) {
+            return dataLayerContent.page[fieldData[PARENT_ID]][DC_TITLE];
+        }
+        return dataLayerContent.component[fieldData[PARENT_ID]][DC_TITLE] ;
+    }
+
+    const getEventInfo = (_eventTitle, _formTitle, _fieldId, dataLayerContent) => {
+        // global variables to be used in abandon event which can be triggered anytime
+        const fieldData = getCurrentFieldData(_fieldId, dataLayerContent);
+        formTitle = _formTitle;
+        fieldTitle = fieldData[DC_TITLE];
+        fieldType = fieldData[FIELD_TYPE];
+        panelTitle = getFieldPanelTitle(fieldData, dataLayerContent);
+        return {
+            formTitle,
+            fieldTitle,
+            fieldType,
+            panelTitle,
+            eventName: _eventTitle
+        }
+    }
+
+    const registerGuideBridgeEvents = () => {
+        window.guideBridge.connect(()=>{
+            const dataLayerContent = dataLayer.getState();
+            formTitle = getFormName(dataLayerContent);
+
+            dataLayer.push({
+                event: FASTTRACK_ANALYTICS_EVENT,
+                eventInfo: {
+                    formTitle: formTitle,
+                    eventName: FormEvents.RENDER
+                }
+            });
+
+            guideBridge.on('elementFocusChanged', function(event){
+                dataLayer.push({
+                    event: FASTTRACK_ANALYTICS_EVENT,
+                    eventInfo: getEventInfo(FormEvents.FIELD, formTitle, event.detail.fieldId, dataLayerContent)
+                });
+            });
+
+            guideBridge.on('elementHelpShown', function(event){
+                dataLayer.push({
+                    event: FASTTRACK_ANALYTICS_EVENT,
+                    eventInfo: getEventInfo(FormEvents.HELP, formTitle, event.detail.fieldId, dataLayerContent)
+                });
+            });
+
+            guideBridge.on('elementErrorShown', function(event){
+                dataLayer.push({
+                    event: FASTTRACK_ANALYTICS_EVENT,
+                    eventInfo: getEventInfo(FormEvents.ERROR, formTitle, event.detail.fieldId, dataLayerContent)
+                });
+            });
+
+            guideBridge.on('submitStart', function(event){
```

Comment:
The form submit doesn't wait for this function to finish, which results in intermittent behavior. Could you please suggest some other way of tracking submit event?

Reply:
it;s just for denoting submitStart (when user clicks on submit button).. form submit should finish after the submit API returns success.
what do you exactly intend to measure? if its time spent on form filling, then it should be correct.

Reply:
Please ignore this, This was a silly mistake on my part after taking the latest code from the dev branch.